File: src/services/childrenService.ts
import { supabase } from '../config/supabase';
import { Child } from '../types/database';

export interface CreateChildData {
  first_name: string;
  last_name: string;
  birth_date: string;
  allergies?: string;
  medical_notes?: string;
  emergency_contact?: string;
  emergency_phone?: string;
}

export interface UpdateChildData extends Partial<CreateChildData> {
  id: string;
}

export class ChildrenService {
  static async getChildren(parentId: string): Promise<Child[]> {
    const { data, error } = await supabase
      .from('children')
      .select('*')
      .eq('parent_id', parentId)
      .order('created_at', { ascending: false });

    if (error) {
      throw new Error(error.message);
    }

    return data || [];
  }

  static async getChildById(childId: string): Promise<Child | null> {
    const { data, error } = await supabase
      .from('children')
      .select('*')
      .eq('id', childId)
      .single();

    if (error) {
      if (error.code === 'PGRST116') {
        return null; // Enfant non trouvé
      }
      throw new Error(error.message);
    }

    return data;
  }

  static async createChild(
    parentId: string,
    childData: CreateChildData
  ): Promise<Child> {
    const { data, error } = await supabase
      .from('children')
      .insert({
        parent_id: parentId,
        ...childData,
        created_at: new Date().toISOString(),
        updated_at: new Date().toISOString(),
      })
      .select()
      .single();

    if (error) {
      throw new Error(error.message);
    }

    return data;
  }

  static async updateChild(childData: UpdateChildData): Promise<Child> {
    const { id, ...updates } = childData;

    const { data, error } = await supabase
      .from('children')
      .update({
        ...updates,
        updated_at: new Date().toISOString(),
      })
      .eq('id', id)
      .select()
      .single();

    if (error) {
      throw new Error(error.message);
    }

    return data;
  }

  static async deleteChild(childId: string): Promise<void> {
    const { error } = await supabase
      .from('children')
      .delete()
      .eq('id', childId);

    if (error) {
      throw new Error(error.message);
    }
  }

  static async getChildrenByNounou(nounouId: string): Promise<Child[]> {
    // Récupérer les enfants via les relations utilisateur
    const { data: relationships, error: relError } = await supabase
      .from('user_relationships')
      .select('parent_id')
      .eq('nounou_id', nounouId)
      .eq('status', 'active');

    if (relError) {
      throw new Error(relError.message);
    }

    if (!relationships || relationships.length === 0) {
      return [];
    }

    const parentIds = relationships.map(rel => rel.parent_id);

    const { data, error } = await supabase
      .from('children')
      .select('*')
      .in('parent_id', parentIds)
      .order('created_at', { ascending: false });

    if (error) {
      throw new Error(error.message);
    }

    return data || [];
  }

  static async searchChildren(
    parentId: string,
    searchTerm: string
  ): Promise<Child[]> {
    const { data, error } = await supabase
      .from('children')
      .select('*')
      .eq('parent_id', parentId)
      .or(`first_name.ilike.%${searchTerm}%,last_name.ilike.%${searchTerm}%`)
      .order('created_at', { ascending: false });

    if (error) {
      throw new Error(error.message);
    }

    return data || [];
  }

  static async getChildrenCount(parentId: string): Promise<number> {
    const { count, error } = await supabase
      .from('children')
      .select('*', { count: 'exact', head: true })
      .eq('parent_id', parentId);

    if (error) {
      throw new Error(error.message);
    }

    return count || 0;
  }

  static async getChildrenWithEvents(
    parentId: string
  ): Promise<(Child & { events_count: number })[]> {
    const { data, error } = await supabase
      .from('children')
      .select(
        `
        *,
        events:events(count)
      `
      )
      .eq('parent_id', parentId)
      .order('created_at', { ascending: false });

    if (error) {
      throw new Error(error.message);
    }

    return (data || []).map(child => ({
      ...child,
      events_count: child.events?.[0]?.count || 0,
    }));
  }
}

File: Downloads/Manounou/flutterflow_export/lib/core/services/children_service.dart
import 'dart:io';
import 'dart:typed_data';
import 'package:flutter/foundation.dart';
import 'package:supabase_flutter/supabase_flutter.dart';
import 'package:image_picker/image_picker.dart';
import 'avatar_service.dart';

class Child {
  final String id;
  final String parentId;
  final String firstName;
  final DateTime? birthDate;
  final String? info;
  final String? photoUrl;
  final String? gender; // 'M' ou 'F'
  final DateTime createdAt;
  final DateTime updatedAt;

  Child({
    required this.id,
    required this.parentId,
    required this.firstName,
    this.birthDate,
    this.info,
    this.photoUrl,
    this.gender,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Child.fromJson(Map<String, dynamic> json) {
    return Child(
      id: json['id'] as String,
      parentId: json['parent_id'] as String,
      firstName: json['first_name'] as String,
      birthDate: json['birth_date'] != null
          ? DateTime.parse(json['birth_date'] as String)
          : null,
      info: json['info'] as String?,
      photoUrl: json['photo_url'] as String?,
      gender: json['gender'] as String?,
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'parent_id': parentId,
      'first_name': firstName,
      'birth_date': birthDate?.toIso8601String().split('T')[0],
      'info': info,
      'photo_url': photoUrl,
      'gender': gender,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
    };
  }

  int? get age {
    if (birthDate == null) return null;
    final now = DateTime.now();
    int age = now.year - birthDate!.year;
    if (now.month < birthDate!.month ||
        (now.month == birthDate!.month && now.day < birthDate!.day)) {
      age--;
    }
    return age;
  }
}

class ChildrenService extends ChangeNotifier {
  final SupabaseClient _supabase = Supabase.instance.client;
  List<Child> _children = [];
  bool _isLoading = false;

  List<Child> get children => _children;
  bool get isLoading => _isLoading;

  // Récupérer tous les enfants de l'utilisateur connecté
  Future<void> loadChildren() async {
    _isLoading = true;
    notifyListeners();

    try {
      // Vérifier la session avant de faire la requête
      final session = _supabase.auth.currentSession;
      if (session == null) {
        debugPrint('⚠️ Aucune session active, redirection vers login recommandée');
        _children = [];
        _isLoading = false;
        notifyListeners();
        return;
      }

      // Vérifier si le token est expiré et essayer de le rafraîchir
      if (session.isExpired) {
        debugPrint('🔄 Session expirée, tentative de rafraîchissement...');
        try {
          await _supabase.auth.refreshSession();
          debugPrint('✅ Session rafraîchie avec succès');
        } catch (refreshError) {
          debugPrint('❌ Erreur lors du rafraîchissement de session: $refreshError');
          // Si le rafraîchissement échoue, vider la liste et laisser l'UI gérer
          _children = [];
          _isLoading = false;
          notifyListeners();
          return;
        }
      }

      final userId = _supabase.auth.currentUser?.id;
      if (userId == null) {
        debugPrint('⚠️ Aucun utilisateur connecté');
        _children = [];
        _isLoading = false;
        notifyListeners();
        return;
      }

      final response = await _supabase
          .from('children')
          .select('*')
          .eq('parent_id', userId)
          .order('created_at', ascending: false);

      _children = (response as List)
          .map((json) => Child.fromJson(json as Map<String, dynamic>))
          .toList();
    } catch (e) {
      debugPrint('❌ Erreur loadChildren: $e');
      
      // Si c'est une erreur d'authentification, vider la liste
      if (e.toString().contains('Auth') || 
          e.toString().contains('session') ||
          e.toString().contains('oauth_client_id')) {
        debugPrint('⚠️ Erreur d\'authentification détectée, vidage de la liste');
        _children = [];
      }
      
      // Ne pas rethrow pour éviter de crasher l'UI
      // L'erreur sera gérée par le dashboard qui affichera un état d'erreur
    } finally {
      _isLoading = false;
      notifyListeners();
    }
  }

  // Upload une photo d'enfant
  // Supporte mobile (File) et Web (XFile avec bytes)
  Future<String?> uploadChildPhoto(String childId, XFile imageFile) async {
    try {
      final userId = _supabase.auth.currentUser?.id;
      if (userId == null) {
        throw Exception('Utilisateur non connecté');
      }

      final fileName = '${DateTime.now().millisecondsSinceEpoch}.jpg';
      final storagePath = '$userId/$childId/$fileName';

      // Sur Web, lire les bytes depuis XFile
      // Sur mobile, utiliser File directement
      Uint8List? bytes;
      if (kIsWeb) {
        bytes = await imageFile.readAsBytes();
      }

      if (kIsWeb && bytes != null) {
        // Upload sur Web avec bytes
        await _supabase.storage
            .from('children-photos')
            .uploadBinary(storagePath, bytes, fileOptions: const FileOptions(
              upsert: true,
              contentType: 'image/jpeg',
            ));
      } else {
        // Upload sur mobile avec File
        final file = File(imageFile.path);
        await _supabase.storage
            .from('children-photos')
            .upload(storagePath, file, fileOptions: const FileOptions(
              upsert: true,
              contentType: 'image/jpeg',
            ));
      }

      // Bucket public : utiliser getPublicUrl()
      final photoUrl = _supabase.storage
          .from('children-photos')
          .getPublicUrl(storagePath);

      return photoUrl;
    } catch (e) {
      debugPrint('Erreur uploadChildPhoto: $e');
      rethrow;
    }
  }

  // Créer un enfant
  Future<Child> createChild({
    required String firstName,
    DateTime? birthDate,
    String? info,
    XFile? photoFile,
    String? gender,
  }) async {
    try {
      final userId = _supabase.auth.currentUser?.id;
      if (userId == null) {
        throw Exception('Utilisateur non connecté');
      }

      // Générer un avatar aléatoire si aucune photo n'est fournie
      String? initialPhotoUrl;
      if (photoFile == null && gender != null) {
        final genderEnum = AvatarService.genderFromString(gender);
        final avatarPath = AvatarService.getRandomAvatarPath(genderEnum);
        // Préfixe "avatar:" pour distinguer les avatars des photos réseau
        initialPhotoUrl = 'avatar:$avatarPath';
      }

      // Créer l'enfant d'abord
      final response = await _supabase.from('children').insert({
        'parent_id': userId,
        'first_name': firstName,
        'birth_date': birthDate?.toIso8601String().split('T')[0],
        'info': info,
        'gender': gender,
        'photo_url': initialPhotoUrl, // Avatar assigné par défaut si pas de photo
      }).select().single();

      var child = Child.fromJson(response);

      // Upload photo si fournie (remplace l'avatar)
      if (photoFile != null) {
        try {
          final photoUrl = await uploadChildPhoto(child.id, photoFile);
          if (photoUrl != null) {
            // Mettre à jour avec l'URL de la photo
            await _supabase
                .from('children')
                .update({'photo_url': photoUrl})
                .eq('id', child.id);
            
            child = Child(
              id: child.id,
              parentId: child.parentId,
              firstName: child.firstName,
              birthDate: child.birthDate,
              info: child.info,
              photoUrl: photoUrl,
              gender: child.gender,
              createdAt: child.createdAt,
              updatedAt: child.updatedAt,
            );
          }
        } catch (e) {
          debugPrint('Erreur upload photo lors création: $e');
          // Continue même si l'upload photo échoue
        }
      }

      _children.insert(0, child);
      notifyListeners();

      return child;
    } catch (e) {
      debugPrint('Erreur createChild: $e');
      rethrow;
    }
  }

  // Mettre à jour un enfant
  Future<void> updateChild({
    required String childId,
    String? firstName,
    DateTime? birthDate,
    String? info,
    XFile? photoFile,
    bool deletePhoto = false,
    String? gender,
  }) async {
    try {
      final updates = <String, dynamic>{};
      if (firstName != null) updates['first_name'] = firstName;
      if (birthDate != null) {
        updates['birth_date'] = birthDate.toIso8601String().split('T')[0];
      }
      if (info != null) updates['info'] = info;
      if (gender != null) {
        updates['gender'] = gender;
        // Si on change le genre et qu'il n'y a pas de photo, générer un nouvel avatar
        if (deletePhoto || (photoFile == null && !updates.containsKey('photo_url'))) {
          final genderEnum = AvatarService.genderFromString(gender);
          final avatarPath = AvatarService.getRandomAvatarPath(genderEnum);
          updates['photo_url'] = 'avatar:$avatarPath';
        }
      }

      // Gérer la photo
      if (deletePhoto) {
        // Si on supprime la photo, générer un nouvel avatar si un genre est disponible
        if (gender != null) {
          final genderEnum = AvatarService.genderFromString(gender);
          if (genderEnum != null) {
            final avatarPath = AvatarService.getRandomAvatarPath(genderEnum);
            updates['photo_url'] = 'avatar:$avatarPath';
          } else {
            updates['photo_url'] = null;
          }
        } else {
          // Récupérer le genre actuel pour générer un avatar
          try {
            final currentChild = await getChildById(childId);
            if (currentChild?.gender != null) {
              final genderEnum = AvatarService.genderFromString(currentChild!.gender);
              if (genderEnum != null) {
                final avatarPath = AvatarService.getRandomAvatarPath(genderEnum);
                updates['photo_url'] = 'avatar:$avatarPath';
              } else {
                updates['photo_url'] = null;
              }
            } else {
              updates['photo_url'] = null;
            }
          } catch (e) {
            debugPrint('Erreur récupération enfant pour avatar: $e');
            updates['photo_url'] = null;
          }
        }
        // TODO: Supprimer le fichier du storage aussi si c'était une vraie photo
      } else if (photoFile != null) {
        // Si une nouvelle photo est fournie, elle remplace toujours l'avatar ou la photo existante
        try {
          final photoUrl = await uploadChildPhoto(childId, photoFile);
          if (photoUrl != null) {
            updates['photo_url'] = photoUrl;
          }
        } catch (e) {
          debugPrint('Erreur upload photo lors mise à jour: $e');
          // Continue même si l'upload photo échoue
        }
      }

      await _supabase
          .from('children')
          .update(updates)
          .eq('id', childId);

      // Recharger la liste
      await loadChildren();
    } catch (e) {
      debugPrint('Erreur updateChild: $e');
      rethrow;
    }
  }

  // Supprimer un enfant
  Future<void> deleteChild(String childId) async {
    try {
      await _supabase.from('children').delete().eq('id', childId);

      _children.removeWhere((child) => child.id == childId);
      notifyListeners();
    } catch (e) {
      debugPrint('Erreur deleteChild: $e');
      rethrow;
    }
  }

  // Récupérer un enfant par ID
  Future<Child?> getChildById(String childId) async {
    try {
      final response = await _supabase
          .from('children')
          .select('*')
          .eq('id', childId)
          .single();

      return Child.fromJson(response);
    } catch (e) {
      debugPrint('Erreur getChildById: $e');
      return null;
    }
  }
}

